transmat/zip/zip_unpack.go

```diff
@@ -0,0 +1,202 @@
+		}
+
+		// Reshuffle metainfo to our default format.
+		skipMe, haltMe := ZipHdrToMetadata(&f.FileHeader, &fmeta)
+		if skipMe != nil {
+			// n.b. every time this happens in practice so far, it's a `g` header,
+			//  and it's got a git commit hash in the paxrecords -- purely advisory info.
+			//  It would be nice to turn those down into a debug message, and react
+			//  a little more startledly to any other values.  Until then, "warn"
+			//  even though every time we've seen this it's harmless.
+			mon.Send(rio.Event_Log{
+				Time:  time.Now(),
+				Level: rio.LogWarn,
+				Msg:   fmt.Sprintf("unpacking: skipping an entry: %s", skipMe),
+				Detail: [][2]string{
+					{"path", fmeta.Name.String()},
+					{"skipreason", skipMe.Error()},
+					{"ziphdr", fmt.Sprintf("%#v", f)},
+				},
+			})
+			continue
+		}
+		if haltMe != nil {
+			return api.WareID{}, api.WareID{}, haltMe
+		}
+		if strings.HasPrefix(fmeta.Name.String(), "..") {
+			return api.WareID{}, api.WareID{}, Errorf(rio.ErrWareCorrupt, "corrupt zip: paths that use '../' to leave the base dir are invalid")
+		}
+
+		// Infer parents, if necessary.  The zip format should not allow implicit dirs, but we allow
+		// it for tars, so why not here.
+		for _, parent := range fmeta.Name.SplitParent() {
+			// If we already initialized this parent, superb; move along.
+			if _, exists := dirs[parent]; exists {
+				continue
+			}
+			// If we're missing a dir, conjure a node with defaulted values.
+			log.DirectoryInferred(mon, parent, fmeta.Name)
+			conjuredFmeta := fshash.DefaultDirMetadata()
+			conjuredFmeta.Name = parent
+			prefilterBucket.AddRecord(conjuredFmeta, nil)
+			filters.ApplyUnpackFilter(filt, &conjuredFmeta)
+			filteredBucket.AddRecord(conjuredFmeta, nil)
+			dirs[conjuredFmeta.Name] = struct{}{}
+			if err := fsOp.PlaceFile(afs, conjuredFmeta, nil, false); err != nil {
+				return api.WareID{}, api.WareID{}, Errorf(rio.ErrInoperablePath, "error while unpacking: %s", err)
+			}
+		}
+
+		// Apply filters.
+		//  ... uck, to one copy of the meta.  We can't add either to their buckets
+		//  until after the file is placed because we need the content hash.
+		filteredFmeta := fmeta
+		//  The filter may reject things by returning an error;
+		//   or, instruct us to ignore things by setting the type to invalid.
+		if err := filters.ApplyUnpackFilter(filt, &filteredFmeta); err != nil {
+			return api.WareID{}, api.WareID{}, err
+		}
+		if fmeta.Type == fs.Type_Invalid {
+			// skip placing that file and continue processing...
+			//  but *do* still record it in the prefilter bucket for hashing.
+			//  (n.b. currently only non-files are ever ejected like this.)
+			prefilterBucket.AddRecord(fmeta, nil)
+			continue
+		}
+
+		// Place the file.
+		switch fmeta.Type {
+		case fs.Type_File:
+			r, err := f.Open()
+			if err != nil {
+				return api.WareID{}, api.WareID{}, Errorf(rio.ErrInoperablePath, "error while unpacking: %s", err)
```

Comment:
I don't think this should be `ErrInoperablePath`... it's probably `ErrWareCorrupt`, isn't it?  `f` being `Open`'d here is a https://golang.org/pkg/archive/zip/#File , not the local file we're creating/unpacking.

(I might consider renaming `f` to `zf` throughout function, to reduce the odds of confusion with a local `os.File` handle.)